ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,160 @@
+(*
+ * Copyright (c) Cloud Software Group, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let inode_size = 4096L
+
+let ( ++ ) = Int64.add
+
+let ( // ) = Filename.concat
+
+type unpack_error =
+  | Illegal_file_path of string
+  | Unsupported_file_type of string
+  | Unpacked_exceeds_max_size_limit of Int64.t
+  | File_size_mismatch of {
+        path: string
+      ; expected_size: Int64.t
+      ; actual_size: Int64.t
+    }
+  | File_incomplete
+  | File_corrupted
+  | Unpacking_failure
+
+let unpack_error_to_string = function
+  | Illegal_file_path path ->
+      Printf.sprintf "Illegal file path: %s" path
+  | Unsupported_file_type t ->
+      Printf.sprintf "Unsupported file type: %s" t
+  | Unpacked_exceeds_max_size_limit size ->
+      Printf.sprintf
+        "Stop unpacking, otherwise unpacked files exceed max size limit: %s"
+        (Int64.to_string size)
+  | File_size_mismatch {path; expected_size; actual_size} ->
+      Printf.sprintf
+        "Unpacked file size mismatch, path: %s, expected size: %s, actual \
+         size: %s"
+        path
+        (Int64.to_string expected_size)
+        (Int64.to_string actual_size)
+  | File_incomplete ->
+      "File incompete"
+  | File_corrupted ->
+      "File corrupted"
+  | Unpacking_failure ->
+      "Unpacking failure"
+
+type unpack_result = Next of Int64.t
+
+let unpack dir max_size_limit acc_size ifd head =
+  let assert_file_path_is_legal file_path =
```

Comment:
Do we have to watch out for file paths like `../../` that escape from the current directory?

Reply:
What do you mean? `../../` has been considered in the function `assert_file_path_is_legal`.

Reply:
Can you explain how  `assert_file_path_is_legal` works? Which path is tested here and how does it detect that a path like `/some/path/../../../../etc/passwd` inside a tar?

Reply:
It will split the path into a dir/file list with `/`. If one of directories is `.` or `..`, it will raise the error. 
Do you mean I should add this case (escape from the current directory) explicitly?

Reply:
Please add a test case that contains a tar file that tries to escape the current directory. I don't see how the code would detect that but you seem to be confident.

Reply:
@contificate Yes, you're right. Let me add some description.
I just thought of another case: if a tar file contains a absulate path, e.g. `/etc/passwd`, which doesn't contain `.` or `..`, it can't be detected by the code. I will also consider about this case.